guides/source/error_reporting.md

```diff
@@ -0,0 +1,160 @@
+
+```rb
+Rails.error.subscribe(ErrorSubscriber.new)
+```
+
+To test the error subscriber, try this in Rails console:
+
+```
+irb(main):001:0> Rails.error.handle { raise }
+```
+
+And see if the error is reported to the service you use.
+
+#### Libraries Subscribe To Rails Reporter
+
+Some libraries may provide their own subscriber classes. Please check their documentation for more information.
+
+- [Sentry](https://sentry.io/) ([document](https://github.com/getsentry/sentry-ruby/blob/master/sentry-rails/lib/sentry/rails/error_subscriber.rb))
+
+### Capture and Report Errors
+
+You can wrap your code inside a block with the reporting APIs, which will report the exceptions surface from the block.
+
+To report and **swallow** the error, use [`Rails.error.handle`](https://api.rubyonrails.org/classes/ActiveSupport/ErrorReporter.html#method-i-handle):
+
+```rb
+Rails.error.handle do
+  1 + '1' # raises TypeError
+end
+1 + 1 # This will be execute
+```
+
+The error will be reported with `handled: true`
+
+To report but **not swallow** the error, use [`Rails.error.record`](https://api.rubyonrails.org/classes/ActiveSupport/ErrorReporter.html#method-i-record):
+
+```rb
+Rails.error.record do
+  1 + '1' # raises TypeError
+end
+1 + 1 # This won't be executed
+```
+
+The error will be reported with `handled: false`
+
+If you decide to rescue the exception manually, you can also report it with [`Rails.error.report`](https://api.rubyonrails.org/classes/ActiveSupport/ErrorReporter.html#method-i-report):
+
+```rb
+begin
+  # code
+rescue StandardError => e
+  Rails.error.report(e)
+end
+```
+
+#### Options
+
+All 3 reporting APIs (`#handle`, `#record`, and `#report`) support the same options:
+
+- `handled`: a `Boolean` to tell if the error was handled
+- `severity`: a `Symbol` about the severity of the exception. Expected values are: `:error`, `:warning`, and `:info`
+- `context`: a `Hash` to provide more context about the error, like request headers or record attributes
+- `source`: a `String` about the source of the exception. Default is `"application"`
+    - You can use it to skip exceptions from certain sources
+
+### Setting Context
+
+In addition to setting context through the `context` option, you can also use the [`#set_context`](https://api.rubyonrails.org/classes/ActiveSupport/ErrorReporter.html#method-i-set_context) API. For example:
+
+```rb
+Rails.error.set_context(section: "checkout", user_id: @user.id)
+```
+
+The context set this way will be merged with the `context` option
+
+```rb
+Rails.error.set_context(a: 1)
+Rails.error.handle(context: { b: 2 }) { raise }
+# the reported context will be: {:a=>1, :b=>2}
+```
+
+### For Libraries
+
+Libraries can easily register their subscribers in `Railtie`:
+
+```rb
+module MySdk
+  class Railtie < ::Rails::Railtie
+    config.after_initialize do |app|
+      app.executor.error_reporter.subscribe(MyErrorSubscriber.new)
+    end
```

Comment:
Should we introduce an API to make this easier? Is there any reason why we need users to know about the executor or even about the `after_initialize`?

Reply:
That's an oversight on my review, this can be simplified to:

```
Rails.error.subscribe(MyErrorSubscriber.new)  
```

From an initializer.

Reply:
We (Sentry) put it in `after_initialize` because we need to wait for the user-defined SDK initializer to finish (usually `config/initializers/sentry.rb`), so we can decided whether to register it based on the user's SDK config.

Reply:
Right, but that's a special case / transitory thing.

Reply:
I think it may be the case to other SDKs too, at least for the near future. Because this is still kinda new and it's better to leave a configuration for user to opt out/in.
We can suggest putting it in an initializer first, but mentioning `after_initialize` is also an option for making it configurable?

Reply:
I expect the people who implement error reporting services gems to know Rails enough to figure this.

Reply:
Ok, fair enough.